doc/mpls/MPLS_hld.md

```diff
@@ -0,0 +1,510 @@
+     sai_get_inseg_entry_attribute_fn               get_inseg_entry_attribute;
+
++    sai_bulk_create_inseg_entry_fn                 create_inseg_entries;
++    sai_bulk_remove_inseg_entry_fn                 remove_inseg_entries;
++    sai_bulk_set_inseg_entry_attribute_fn          set_inseg_entries_attribute;
++    sai_bulk_get_inseg_entry_attribute_fn          get_inseg_entries_attribute;
+
+} sai_mpls_api_t;
+```
+
+#### Next Hop
+Full details about SAI Next Hop API and attributes are described here:
+https://github.com/opencomputeproject/SAI/blob/master/inc/sainexthop.h
+
+No modifications were made to the current SAI Next Hop API definition.
+
+The following existing attribute is now introduced to SONiC orchagent to facilitate MPLS NextHop functionality:
+```
+    /**
+     * @brief Push label
+     *
+     * @type sai_u32_list_t
+     * @flags MANDATORY_ON_CREATE | CREATE_ONLY
+     * @condition SAI_NEXT_HOP_ATTR_TYPE == SAI_NEXT_HOP_TYPE_MPLS
+     */
+    SAI_NEXT_HOP_ATTR_LABELSTACK,
+```
+
+### Configuration and management
+This section should have sub-sections for all types of configuration and management related design. Example sub-sections for "CLI" and "Config DB" are given below. Sub-sections related to data models (YANG, REST, gNMI, etc.,) should be added as required.
+
+#### CLI Enhancements
+A new SONiC CLI command is introduced to configure interfaces for MPLS.
+
+    # Enable/disable MPLS per INTERFACE.
+    config interface mpls add|remove <intf-name>
+
+#### Config DB Enhancements
+
+##### INTERFACE
+The existing INTERFACE table is enhanced to accept a new "mpls" enable/disable attribute.
+
+``` rfc5234
+INTERFACE|{{ifname}}
+    "mpls":{{enable|disable}} (OPTIONAL)
+
+; Defines schema for MPLS configuration attribute
+key                      = INTERFACE:ifname    ; Interface name
+; value annotations
+ifname                   = 1*64VCHAR           ; name of the Interface
+; field                  = value
+mpls                     = "enable"/"disable"  ; Enable/disable MPLS function. Default is disable
+```
+
+##### PORTCHANNEL_INTERFACE
+The existing PORTCHANNEL_INTERFACE table is enhanced to accept a new "mpls" enable/disable attribute.
+
+``` rfc5234
+PORTCHANNEL_INTERFACE|{{ifname}}
+    "mpls":{{enable|disable}} (OPTIONAL)
+
+; Defines schema for MPLS configuration attributes
+key                      = PORTCHANNEL_INTERFACE:ifname   ; Port Channel Interface name
+;value annotations
+ifname                   = 1*64VCHAR                      ; name of the Interface (Port Channel)
+; field                  = value
+mpls                     = "enable"/"disable"             ; Enable/disable MPLS function. Default disable
+```
+
+### Warmboot and Fastboot Design Impact
+MPLS design will not affect warmboot or fastboot design.
+
+### Restrictions/Limitations
+In this document, MPLS support is only for static LSP route support. The scope of routing stack supporting dynamic creation of MPLS tunnel is not in the design.
```

Comment:
Any plan to add SONIC CLI commands for configuring the MPLS static LSPs?

Reply:
Not in the scope of this implementation yet.